server/src/main/scala/org/scassandra/server/actors/QueryHandler.scala

```diff
@@ -15,37 +15,67 @@
+import org.scassandra.server.actors.ActivityLogActor.RecordQuery
+import org.scassandra.server.actors.ProtocolActor._
+import org.scassandra.server.actors.priming.PrimeQueryStoreActor.{MatchPrime, MatchResult, Reply}
 
-class QueryHandler(primeQueryStore: PrimeQueryStore, activityLog: ActivityLog) extends ProtocolActor {
+import scala.concurrent.duration._
+
+class QueryHandler(primeQueryStore: ActorRef, activityLog: ActorRef) extends ProtocolActor {
+
+  import context.dispatcher
 
   val noRows = Some(Reply(NoRows))
 
-  def receive = {
+  def receive: Receive = {
     case ProtocolMessage(Frame(header, query: Query)) =>
-      implicit val protocolVersion = header.version.version
-      val prime = primeQueryStore(query)
-
-      // Attempt to extract values if variable types are present in the prime.
-      val typesAndValues: Option[(List[DataType], List[Any])] = prime
-        .flatMap(_.variableTypes)
-        .flatMap { variableTypes =>
-          Some(variableTypes).zip(extractQueryVariables(query.query, query.parameters.values.map(_.map(_.value)), variableTypes)).headOption
-        }
+      implicit val protocolVersion: ProtocolVersion = header.version.version
+      implicit val timeout: Timeout = Timeout(1 second)
 
-      typesAndValues match {
-        case Some((variableTypes, values)) =>
-          activityLog.recordQuery(query.query, query.parameters.consistency, query.parameters.serialConsistency, values,
-            variableTypes, query.parameters.timestamp)
-        case None =>
-          activityLog.recordQuery(query.query, query.parameters.consistency, query.parameters.serialConsistency,
-            timestamp = query.parameters.timestamp)
-      }
-
-      writePrime(query, prime, header, alternative = noRows, consistency = Some(query.parameters.consistency))
       log.info(s"Incoming query: $query")
+      val toReply = sender()
+
+      (primeQueryStore ? MatchPrime(query))
+        .mapTo[MatchResult].map { mr =>
+
+        val typesAndValues: Option[(List[DataType], List[Any])] = (for {
+          prime <- mr.prime
+          vt <- prime.variableTypes
+        } yield Some(vt).zip(extractQueryVariables(query.query, query.parameters.values.map(_.map(_.value)), vt)).headOption).flatten
+
+        typesAndValues match {
+          case Some((vts, values)) =>
+            activityLog ! RecordQuery(Activity.Query(query.query, query.parameters.consistency, query.parameters.serialConsistency, values,
+              vts, query.parameters.timestamp))
+          case None =>
+            activityLog ! RecordQuery(Activity.Query(query.query, query.parameters.consistency, query.parameters.serialConsistency,
+              timestamp = query.parameters.timestamp))
+        }
+        writePrime(query, mr.prime, header, alternative = noRows, consistency = Some(query.parameters.consistency), target = toReply)(context.system)
+      }
   }
+
+
+  //      val prime = primeQueryStore(query)
```

Comment:
These comments can probably be removed (unless there for future ref)

Reply:
Nope, this was a mistake, removed